client/peerset/src/protocol_controller.rs

```diff
@@ -0,0 +1,443 @@
+		}
+
+		match self.nodes.entry(peer_id) {
+			Entry::Occupied(mut node) => {
+				if !node.get().is_connected() {
+					return
+				}
+
+				if self.reserved_only {
+					// If we are in reserved-only mode, the removed reserved node
+					// should be disconnected.
+					*node.get_mut() = PeerState::NotConnected;
+					let _ = self.to_notifications.unbounded_send(Message::Drop {
+						set_id: self.set_id,
+						peer_id: *node.key(),
+					});
+					info!(
+						target: "peerset",
+						"Disconnecting previously reserved node {} on {:?}.",
+						node.key(), self.set_id
+					);
+				} else {
+					// Otherwise, just count occupied slots for the node.
+					match node.get() {
+						PeerState::Connected(Direction::Inbound) => self.num_in += 1,
+						PeerState::Connected(Direction::Outbound) => self.num_out += 1,
+						PeerState::NotConnected => {},
+					}
+				}
+			},
+			Entry::Vacant(_) => {
+				debug_assert!(false, "Reserved node not in the list of known nodes.");
+				error!(
+					target: "peerset",
+					"Invalid state: reserved node {} not in the list of known nodes.",
+					peer_id,
+				);
+			},
+		}
+	}
+
+	fn on_set_reserved_peers(&mut self, peer_ids: HashSet<PeerId>) {
+		// Determine the difference between the current group and the new list.
+		let to_insert = peer_ids.difference(&self.reserved_nodes).cloned().collect::<Vec<_>>();
+		let to_remove = self.reserved_nodes.difference(&peer_ids).cloned().collect::<Vec<_>>();
+
+		for node in to_insert {
+			self.on_add_reserved_peer(node);
+		}
+
+		for node in to_remove {
+			self.on_remove_reserved_peer(node);
+		}
+	}
+
+	fn on_set_reserved_only(&mut self, reserved_only: bool) {
+		self.reserved_only = reserved_only;
+
+		if reserved_only {
+			// Disconnect all non-reserved peers.
+			for (peer_id, state) in self.nodes.iter_mut() {
+				match state {
+					PeerState::Connected(d) => {
+						match d {
+							Direction::Inbound => self.num_in -= 1,
+							Direction::Outbound => self.num_out -= 1,
+						}
+						*state = PeerState::NotConnected;
+						let _ = self.to_notifications.unbounded_send(Message::Drop {
+							set_id: self.set_id,
+							peer_id: *peer_id,
+						});
+					},
+					PeerState::NotConnected => {},
+				}
+			}
+		} else {
+			// Try connecting to non-reserved peers.
+			self.alloc_slots();
+		}
+	}
+
+	fn on_add_peer(&mut self, peer_id: PeerId) {
+		if self.nodes.insert(peer_id, PeerState::NotConnected).is_none() {
+			self.alloc_slots();
+		}
+	}
+
+	fn on_remove_peer(&mut self, peer_id: PeerId) {
+		// Don't do anything if the node is reserved.
+		if self.reserved_nodes.contains(&peer_id) {
+			return
+		}
+
+		match self.nodes.remove(&peer_id) {
+			Some(state) => match state {
+				PeerState::Connected(d) => {
+					match d {
+						Direction::Inbound => self.num_in -= 1,
+						Direction::Outbound => self.num_out -= 1,
+					}
+					let _ = self
+						.to_notifications
+						.unbounded_send(Message::Drop { set_id: self.set_id, peer_id });
```

Comment:
I think at least this could be refactored into a separate function since it's being repeated

Reply:
I tried extracting this into a separate function, but the annoying pattern I'm repeatedly hitting here is that after borrowing a hash map entry / value I can't call other methods of `self`, because `self` is already borrowed as mutable. The trick with `Peer` abstraction worked, because it didn't actually held a reference to a map element, but just a `PeerId`.